File: scripts/dmzMBRAPluginCreateObject.lua
function receive (self, type, data)

   if dmz.data.is_a (data) then

      local pos = data:lookup_vector ("position", 1)

      local undoHandle = dmz.undo.start_record ("Create Node")
      local handle = dmz.object.create ("na_node")

      if not handle then self.log:error ("Object Not Created!")
      else
         dmz.object.position (handle, nil, pos)
         local outData = dmz.data.new ()
         outData:store_handle ("object", 1, handle)
         outData:store_handle ("created", 1, handle)
         self.editMessage:send ("NetworkAnalysisNodeProperties", outData)
      end
      if dmz.object.is_object (handle) then
         dmz.object.activate (handle)
         dmz.undo.stop_record (undoHandle)
      else dmz.undo.abort_record (undoHandle)
      end
   end
end

function new (config, name)

   local self = {
      log = dmz.log.new ("lua." .. name),
      message = config:to_message ("message.name", "CreateObjectMessage"),
      editMessage =
         config:to_message ("edit.name", "EditObjectAttributesMessage"),
      obs = dmz.message_observer.new (name),
   }

   self.log:info ("Creating plugin: " .. name)
   self.log:info ("Message type: " .. self.message:get_name ())
   self.obs:register (self.message, receive, self)

   return self
end

File: scripts/dmzMBRAPluginCreateObject.js
var dmz =
      { object: require("dmz/components/object")
      , objectType: require("dmz/runtime/objectType")
      , data: require("dmz/runtime/data")
      , defs: require("dmz/runtime/definitions")
      , message: require("dmz/runtime/messaging")
      , undo: require("dmz/runtime/undo")
      }
   , message = dmz.message.create(
      self.config.string("message.name","CreateObjectMessage"))
   , editMessage = dmz.message.create(
      self.config.string("edit.name","EditObjectAttributesMessage"))
   ;

// Create a new node, place it, then send a message to open the Edit NA node window
// Once the window has been closed successfully, activate the node. If the user hit
// cancel, the node will not be activated.

// NOTE:
// Necessary to convert strings into new named handles before sending messages

message.subscribe(self, function receive (data) {

   var pos
     , undoHandle
     , handle
     , outData
     , nanpHandle
     ;
   if (dmz.data.isTypeOf (data)) {

      pos = data.vector ("position", 0);
      undoHandle = dmz.undo.startRecord ("Create Node");
      handle = dmz.object.create("na_node");

      if (!handle) { self.log.error ("Object Not Created!"); }
      else {
         dmz.object.position (handle, null, pos);
         outData = dmz.data.create ();
         outData.handle ("object", 0, handle);
         outData.handle ("created", 0, handle);
         nanpHandle = dmz.defs.createNamedHandle("NetworkAnalysisNodeProperties");
         editMessage.send (nanpHandle, outData);
      }

      if (dmz.object.isObject (handle)) {
         dmz.object.activate (handle);
         dmz.undo.stopRecord (undoHandle);
      } else { dmz.undo.abortRecord (undoHandle); }
   }
});
